### PR TITLE
PostgreSQL - new sources & actions

### PR DESCRIPTION
**Pull request by michelle0927**
New event sources and new actions for PostgreSQL. Addresses issues https://github.com/PipedreamHQ/pipedream/issues/1034 and https://github.com/PipedreamHQ/pipedream/issues/1483

### CODE REVIEW

**Comment by dylburger** [comment]
- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling soures, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/postgresql/actions/execute-custom-query/execute-custom-query.mjs
Comment (jcortes) [inline components/postgresql/actions/execute-custom-query/execute-custom-query.mjs:19]: Not sure if this will be secured against sql injection, since is not using any kind of [prepared statements](https://node-postgres.com/features/queries#prepared-statements) here. I would suggest to add support for prepared statemens so if you want you can take a look at this example with [MySql execute-query.mjs](https://github.com/PipedreamHQ/pipedream/blob/ac90bf536ce0e911ac0bc7b07d6ba2e23b09e18e/components/mysql/actions/execute-query/execute-query.mjs#L52)

File: components/postgresql/actions/new-row/new-row.mjs
Comment (jcortes) [inline components/postgresql/actions/new-row/new-row.mjs:33]: You should consider the case when a user type an unknown column name
```suggestion
   try {
      const res = await this.postgresql.insertRow(table, columns, values);
      $.export("$summary", "New row inserted");
      return res;

    } catch (error) {
      $.export("$summary", "New row not inserted due to an error");
    }
```

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:36]: I'm not sure about getting all values from one table in this way because the table might have a huge number of records, so pagination would be a must here I guess. The other way I see it is by doing it with just a simple where string (e.g. [MySql delete-row](https://github.com/PipedreamHQ/pipedream/blob/ac90bf536ce0e911ac0bc7b07d6ba2e23b09e18e/components/mysql/actions/delete-row/delete-row.mjs#L27))

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:41]: I guess here we can also give a clue to the user on how to add the key/value pairs when structure mode is off describing the value as `{{{columnName:"columnValue"}}}`?

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:64]:
```suggestion
      return client.end();
```

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:75]: I think you might consider getting errors running the query and if that happens you are not ending the connection so this is my suggestion:
```suggestion
      try {
        const { rows } = await client.query(query);
        return rows;
      } finally {
        await this.endClient(client);
      }
```

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:115]:
```suggestion
      const rows = await this.executeQuery(query);
```

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:155]: You are missing the `await` here. Also you can remove the `.join()` function from the template literal since the array will be converted automatically into an string of names separated by comas.

Also there is a caveat on the sql query syntax here, because it seems like if I'm sending a column name that was declared as a VARCHAR it throws an error. So my suggestion is to map the values and force them to be enclosed by single quotes like this
```javascript
${values.map((v) => "'" + v + "'")
```
Otherwise you need to advice the user in the property description that he needs to set those values inside `''`

File: components/postgresql/sources/new-row/new-row.mjs
Comment (jcortes) [inline components/postgresql/sources/new-row/new-row.mjs:59]: There is an issue with this one because if I have a table with these columns `(id, name)` and I select the column `name` in the setup config, that means the rows will be order by `name`. So then let's supose I'm adding the first row as
```sql
INSERT INTO test_table (id, name) values (1, 'George');
```
and right after the insert I run the source, so in this case the `lastResult` values will be `George`. So then if I need to insert the next row like this
```sql
INSERT INTO test_table (id, name) values (2, 'Andrew');
```
and then the trigger gets called. It won't display this new record since the latest `name` inserted was `George`  and according to the sql statement here in `getRows` function
```sql
...
WHERE " + column + ">" + lastResult
...
```
it won't show the row with `Andrew` in the `name`.

So I think a practical solution for this is when we are deploying we should validate if the colum is `unique` if not, throws with error advicing this to the user. In that way the data would be more consistent when it comes to get new inserted rows. I was testing this on Zapier and it's the same aproach, they seem to not allow to select a column that is not unique before proceed with next step.

Another solution would be to make the user know that the `column` property should be datetime type the same aproach that you took for  `new-or-updated-row.mjs` in the property description.

Let me see what are your thoughts on this.

**Review by jcortes — CHANGES_REQUESTED** [review]
I've just left a few comments/suggestions but overall looks good to me!

File: components/postgresql/postgresql.app.mjs
Comment (michelle0927) [reply components/postgresql/postgresql.app.mjs]: Updated to use pagination.

File: components/postgresql/sources/new-row/new-row.mjs
Comment (michelle0927) [reply components/postgresql/sources/new-row/new-row.mjs]: I updated `new-row.mjs` so that it doesn't use `lastResult`. It will just use the selected column to dedupe. I also added in a method to check that the column provided is unique (checks that it contains no duplicate values) which is used in sources `new-or-updated-row.mjs`, `new-row.mjs`, and `new-row-custom-query.mjs`.

File: components/postgresql/actions/delete-rows/delete-rows.mjs
Comment (jcortes) [inline components/postgresql/actions/delete-rows/delete-rows.mjs:48]: So here you can omit both `getRowCount` if you use `RETURNING *` in the delete sql statement

File: components/postgresql/actions/execute-custom-query/execute-custom-query.mjs
Comment (jcortes) [inline components/postgresql/actions/execute-custom-query/execute-custom-query.mjs:39]: If you change `executeQuery` as I suggested in `postgresql.app.mjs` I would refactor this one to
```suggestion
    const res = await this.postgresql.executeQuery({
      text: query,
      values,
    });
```

File: components/postgresql/actions/find-row-custom-query/find-row-custom-query.mjs
Comment (jcortes) [inline components/postgresql/actions/find-row-custom-query/find-row-custom-query.mjs:3]: This particular component is kind of confusing becase the name suggests it's only for finding rows with a custom query but at the same time the query is too open to even run an insert/update/delete as well, so I think this one could easily be replaced by `execute-custom-query` don´t you think so?

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:1]: I would also suggest using [pg-format lib](https://www.npmjs.com/package/pg-format) to also prevent potential SQL Injection. It is mentioned [here as well](https://node-postgres.com/features/queries#parameterized-query)

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:98]: As a suggestion I would leave the `query` object to be configurable outside of this function to make it available to other functions. It migth be helpful in case you want to add more properties to `query`.

```suggestion
    async executeQuery(query) {
      const client = await this.getClient();
      try {
        const { rows } = await client.query(query);
        return rows;
      } finally {
        await this.endClient(client);
      }
    },
```

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:113]: If you change `executeQuery` the way I suggested previously I would refactor this one to
```suggestion
      const rows = await this.executeQuery({
        text: `
          SELECT column_name FROM INFORMATION_SCHEMA.COLUMNS
            WHERE TABLE_NAME = $1
        `,
        values: [
          table,
        ],
      });
```

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:122]: If you change `executeQuery` the way I suggested previously I would refactor this one to
```suggestion
      const rows = await this.executeQuery({
        text: `
          SELECT a.attname FROM pg_index i
            JOIN pg_attribute a ON a.attrelid = i.indrelid
              AND a.attnum = ANY(i.indkey)
            WHERE i.indrelid = $1::regclass
              AND i.indisprimary
        `,
        values: [
          table,
        ],
      });
```

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:137]:
```suggestion
      const select = "SELECT * FROM %I";
      const where = "WHERE %I > $1";
      const orderby = "ORDER BY %I DESC";

      const query = lastResult
        ? {
          text: format(`${select} ${where} ${orderby}`, table, column, column),
          values: [
            lastResult,
          ],
        }
        : format(`${select} ${orderby}`, table, column);

      return this.executeQuery(query);
```

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:147]: If you change `executeQuery` the way I suggested previously I would refactor this one to
```suggestion
      const rows = await this.executeQuery({
        text: format("SELECT * FROM %I WHERE %I = $1", table, column),
        values: [
          value,
        ],
      });
```

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:158]: I think this function won´t be necessary if you get the array of rows affected by `RETURNING *` in the delete sql statement.

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:167]: If you change `executeQuery` the way I suggested previously I would refactor this one to
```suggestion
   async deleteRows(table, column, value) {
      return this.executeQuery({
        text: format("DELETE FROM %I WHERE %I = $1 RETURNING *", table, column),
        values: [
          value,
        ],
      });
    },
```
Also with this aproach you won´t need to call twice the function `getRowCount` in order to get the number of deleted rows

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:177]: If you change `executeQuery` the way I suggested previously I would refactor this one to
```suggestion
    async insertRow(table, columns, values) {
      const placeholders = this.getPlaceholders({
        values,
      });
      return this.executeQuery({
        text: format(`
          INSERT INTO %I (${columns}) 
            VALUES (${placeholders})
            RETURNING *
        `, table),
        values,
      });
    },
```

with this additional function which will give you the necessary placeholder values for the query
```javascript
    getPlaceholders({
      values = [], fromIndex = 1,
    }) {
      return values.map((_, index) => `$${index + fromIndex}`);
    },
```

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:201]: If you change `executeQuery` the way I suggested previously I would refactor this one to

```suggestion
    async updateRow(table, lookupColumn, lookupValue, rowValues) {
      const primaryKey = await this.getPrimaryKey(table);
      const columnsPlaceholders = this.getColumnsPlaceholders({
        rowValues,
        fromIndex: 2,
      });
      const response = await this.executeQuery({
        text: format(`
          UPDATE %I SET ${columnsPlaceholders}
            WHERE %I = (
              SELECT %I FROM %I
              WHERE %I = $1
              ORDER BY %I LIMIT 1
            ) RETURNING *
        `, table, primaryKey, primaryKey, table, lookupColumn, primaryKey),
        values: [
          lookupValue,
          ...Object.values(rowValues),
        ],
      });
      return response[0];
    },
```

with this additional function which will give you the necessary placeholder values for the query
```javascript
    getColumnsPlaceholders({
      rowValues = {}, fromIndex = 1,
    }) {
      return Object.keys(rowValues)
        .map((key, index) => `${key} = $${index + fromIndex}`);
    },
```

File: components/postgresql/postgresql.app.mjs
Comment (jcortes) [inline components/postgresql/postgresql.app.mjs:211]: If you change `executeQuery` the way I suggested previously I would refactor this one to
```suggestion
      const rows = await this.executeQuery({
        text: format(`
          SELECT %I FROM %I
            LIMIT $1::numeric OFFSET $2::numeric
        `, column, table),
        values: [
          limit,
          offset,
        ],
      });
```

File: components/postgresql/sources/common.mjs
Comment (jcortes) [inline components/postgresql/sources/common.mjs:63]: This sql can also be used with `pg-format`

File: components/postgresql/sources/new-row-custom-query/new-row-custom-query.mjs
Comment (jcortes) [inline components/postgresql/sources/new-row-custom-query/new-row-custom-query.mjs:73]: If you change `executeQuery` as I suggested in `postgresql.app.mjs` then you can change this to
```suggestion
    const rows = await this.postgresql.executeQuery({
      text: query,
      values,
    });
```

**Review by jcortes — CHANGES_REQUESTED** [review]
Thanks for making all those changes. There si a few more comments request changes I would like you to address and that would be good enough. Thanks again!

File: components/postgresql/actions/find-row-custom-query/find-row-custom-query.mjs
Comment (michelle0927) [reply components/postgresql/actions/find-row-custom-query/find-row-custom-query.mjs]: I agree. `find-row-custom query` does the same thing as `execute-custom-query` but only returns the first result. I removed this component.

File: components/postgresql/sources/common.mjs
Comment (jcortes) [inline components/postgresql/sources/common.mjs:65]: Optionally you can also set query as a plain string by using `format`
```suggestion
      const query = format("select count(*) <> count(distinct %I) as duplicate_flag from %I", column, table);
```

**Review by jcortes — APPROVED** [review]
Thanks a lot for all those changes again, nice work!

**Comment by dmathukorala** [comment]
Verification completed, issue observed in Update Row Action - Error: Relation Does not exist (Please find the screenshot attached)
image

**Comment by michelle0927** [comment]
@dmathukorala The error was coming from the method `getPrimaryKey()` where the query was using a case insensitive version of the `table` name. I made an update that fixed the issue for me, so let me know if you're still seeing that error.

**Comment by DilanAthukorala** [comment]
@michelle0927 Now observing different error for update_row action
image

**Comment by michelle0927** [comment]
@DilanAthukorala I'm able to reproduce this issue when the table selected doesn't have a primary key constraint. I made an update to `updateRow()` to use the `lookupColumn` instead of the `primaryKey` so that this component should work in tables without a primary key.

**Comment by DilanAthukorala** [comment]
Thanks @michelle0927 this works fine now. Could you also check **New or Updated Row** source I'm not getting any events after updating rows or adding new rows.

**Comment by michelle0927** [comment]
> Thanks @michelle0927 this works fine now. Could you also check **New or Updated Row** source I'm not getting any events after updating rows or adding new rows.

@DilanAthukorala Is the `column` prop that you're entering "A datetime column, such as 'date_updated' or 'last_modified' that is set to the current datetime when a row is updated"?

**Comment by michelle0927** [comment]
@dannyroosevelt This needs an approval before I can merge.

**Review by dannyroosevelt — APPROVED** [review]

**Comment by dannyroosevelt** [comment]
> @dannyroosevelt This needs an approval before I can merge.

approved!

**Pull request merged**
